src/main/java/com/palantir/gradle/versions/VersionsPropsIdeaPlugin.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.gradle.versions;
+
+import java.io.File;
+import java.util.List;
+import java.util.stream.Collectors;
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.transform.OutputKeys;
+import javax.xml.transform.Transformer;
+import javax.xml.transform.TransformerException;
+import javax.xml.transform.TransformerFactory;
+import javax.xml.transform.dom.DOMSource;
+import javax.xml.transform.stream.StreamResult;
+import org.gradle.api.Plugin;
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.repositories.MavenArtifactRepository;
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+
+public final class VersionsPropsIdeaPlugin implements Plugin<Project> {
+
+    @Override
+    public void apply(Project target) {
+        target.getPluginManager().withPlugin("idea", _ideaPlugin -> {
+            configureIntellij(target);
+        });
+    }
+
+    private static void configureIntellij(Project target) {
+        if (!Boolean.getBoolean("idea.active")) {
+            return;
+        }
+        List<String> repositoryUrls = target.getRootProject().getAllprojects().stream()
+                .flatMap(project -> project.getRepositories().withType(MavenArtifactRepository.class).stream())
+                .map(MavenArtifactRepository::getUrl)
+                .map(Object::toString)
+                .map(url -> url.endsWith("/") ? url : url + "/")
+                .distinct()
+                .collect(Collectors.toList());
+
+        File ideaDir = new File(target.getRootProject().getProjectDir(), ".idea");
+
+        writeRepositoryUrlsToXml(repositoryUrls, new File(ideaDir, "mavenRepositories.xml"));
+    }
+
+    private static void writeRepositoryUrlsToXml(List<String> repositoryUrls, File file) {
```

Comment:
It's not immediately clear what the structure of said repository file is without having to run this code in my head. We typically use Jackson (has xml support) together with [immutables](https://immutables.github.io/) (has nice builders). Depending on whether or not you want to couple the apis, you can even use the same object to deserialise it within IntelliJ!

You can do something like:

```gradle
dependencies {
  // assuming you have usual jackson dependencies already
  implementation 'com.fasterxml.jackson.dataformat:jackson-dataformat-xml'
}
```

```java
@Value.Immutable
@JsonDeserialize(as = ImmutableMavenRepositories.class)
@JsonSerialize(as = ImmutableMavenRepositories.class)
@JsonRootName("repositories")
interface MavenRepositories {
  @Value.Parameter
  @JacksonXmlElementWrapper(useWrapping = false)
  @JacksonXmlProperty(localName = "repository")
  List<MavenRepository> repository();
}

@Value.Immutable
@JsonDeserialize(as = ImmutableMavenRepository.class)
@JsonSerialize(as = ImmutableMavenRepository.class)
interface MavenRepository {
  @Value.Parameter
  @JacksonXmlProperty(isAttribute = true)
  String value();
}
```

So to actually do what you want to do:
```java
// make all the objects
Set<String> repositoryUrls = ...; 
// we can actually just combine it here, but I don't want to copy for brevity sake

repositoryUrls.stream()
  .map(ImmutableMavenRepository::of)
  .collect(Collectors.collectingAndThen(Collectors.toList(), ImmutableMavenRepositories::of);
```

In case you're wondering what [`@Value.Parameter`](https://immutables.github.io/immutable.html#constructor-method) does.

It's not strictly necessary, it'd just be a bit more cumbersome:

```java
public static MavenRepository of(String url) {
  return ImmutableMavenRepository.builder()
    .value(url)
    .build();
}
```

## How to write this out?
The above annotations:
* `@JsonDeserialise`
* `@JsonSerialise`
* `@JsonRootName`
* `@JacksonXmlProperty`

are annotations that are part of jackson. Jackson is the serialisation/deserialisation library for many different formats i.e. json, yaml, and in this case xml. They're controlling what jackson is supposed to do on both serialisation and deserialisation.

Once you have an object, you need to serialise it:

```java
XmlMapper mapper = new XmlMapper();

// typically we use ObjectMapper for both yaml and json, 
// but XmlMapper is needed here for xml related things
MavenRepositories mavenRepositories = ...;
mapper.writerWithDefaultPrettyPrinter().writeValue(file, mavenRepositories);
```

If you have an xml file and you want it read it into a nice to use object without having to go through manual document and element apis, you can use the opposite apis:

```java
XmlMapper mapper = new XmlMapper();
MavenRepositories mavenRepositories = mapper.readValue(file, MavenRepositories.class);
```

Yes, the above is more code, but it's easy to refer to what a document might look like without having to run it. It also means you can do things via higher level operations i.e. make an object, and write it out as opposed to make the  right invocations in the right order for the low-level XML writer.

Reply:
thanks for the review this is super detailed! I have now switched to jackson :)